agent/src/Ingest.php

```diff
@@ -83,32 +86,46 @@ private function ingest(string $payload): void
         }
 
         $this->concurrentRequests++;
+        $start = microtime(true);
+
+        $this->ingestDetails->get()->then(function (?IngestDetails $ingestDetails) use ($payload, &$start): PromiseInterface {
+            $start = microtime(true);
 
-        $this->ingestDetails->get()->then(function (?IngestDetails $ingestDetails) use ($payload): PromiseInterface {
             if ($ingestDetails === null) {
-                throw new RuntimeException('Unable to ingest payload. No authentication details found.');
+                throw new RuntimeException('No authentication details');
             }
 
-            $start = microtime(true);
-
             return $this->browser->post(
                 url: $ingestDetails->ingestUrl,
                 headers: [
                     'authorization' => "Bearer {$ingestDetails->token}",
                 ],
                 body: $payload,
-            )->then(
-                function (ResponseInterface $response) use ($start): void {
-                    call_user_func($this->onIngestSuccess, $response, microtime(true) - $start);
-                },
-                function (Throwable $e) use ($start): void {
-                    call_user_func($this->onIngestError, $e, microtime(true) - $start);
-                }
             );
-        })->catch(function (Throwable $e): void {
-            call_user_func($this->onIngestError, $e, 0.0);
```

Comment:
I was able to collapse all the success and error handling down into the parent promise's `then` and `catch` calls rather than having these nested.